add recurrent neural network

JesusPinedaC opened the pull request:
This pull request introduces recurrent neural networks based on @Henrik-KM 's implementation in https://github.com/DeepTrackAI/deeplay/pull/23

## Example:

```python
rnn = dl.RecurrentNeuralNetwork(
    in_features=2,
    hidden_features=[32, 32],
    out_features=10
)
```
Every component of the network is fully flexible. For instance:

```python
rnn.blocks.layer.configure(nn.LSTM)
```

Or, in the case of bidirectional RNNs:

```python
rnn.blocks.layer.configure(bidirectional=True)

# The forward and backward directions are concatenated by the RNN layer, so the input_size of the hidden layers needs to be updated
rnn.blocks[1:].layer.configure(input_size=64)
```

@Henrik-KM, the idea is updating this implementation based on your inputs!

JesusPinedaC committed:
add recurrent neural network

JesusPinedaC committed:
Update rnn.py

JesusPinedaC committed:
Create test_rnn.py

Henrik-KM committed:
Added models package, with first model RNN, and LAND layers with custom interlayer dropout

Henrik-KM committed:
Revert "Added models package, with first model RNN, and LAND layers with custom interlayer dropout"

This reverts commit 54698b702e3de9e95e9260d077736ba7c78971ff.

Henrik-KM committed:
Jp/rnns
Added models package, with first model RNN, and LAND layers with custominterlayer dropout

Henrik-KM committed:
Revert "Jp/rnns"

This reverts commit 2f55e61b6d05d015e10b861dfe37b6870a9654be.

Henrik-KM committed:
Jp/rnns
Added models, interlayer dropout, RNN model

Henrik-KM committed:
Jp/rnns
Added insert command to list, added interlayer dropout to MLP component

giovannivolpe committed:
Merge pull request #44 from Henrik-KM/jp/rnns

Jp/rnns

giovannivolpe commented:
@BenjaminMidtvedt @Henrik-KM there are some conflicts that need to be resolved before merging this. Can you check and confirm we can proceed to merge?

BenjaminMidtvedt commented on deeplay/models/RNN.py:
```diff
@@ -0,0 +1,225 @@
+        if embedding:
+            self.embedding_dropout=torch.nn.Dropout(dropout)
+
+        if self.rnn_type == 'LSTM':
+            self.rnn_class = torch.nn.LSTM
+        elif self.rnn_type == 'GRU':
+            self.rnn_class = torch.nn.GRU
+        else:
+            self.rnn_class = torch.nn.RNN
```
does not handle case where rnn_type is a module.

BenjaminMidtvedt commented on deeplay/models/RNN.py:
```diff
@@ -0,0 +1,225 @@
+        in_features: Optional[int], 
+        hidden_features: Sequence[int], 
+        out_features: Optional[int] = None,
+        rnn_type: Union[Literal['RNN', 'LSTM', 'GRU'],torch.nn.Module] = 'GRU',
```
should probably be typed Type[torch.nn.Module]

BenjaminMidtvedt commented on deeplay/blocks/LAND.py:
rename lower case land

BenjaminMidtvedt commented on deeplay/list.py:
```diff
@@ -37,6 +37,18 @@ def append(self, module: DeeplayModule) -> "LayerList[T]":
 
         return self
+    
+    def insert(self, index: int, module: DeeplayModule) -> "LayerList[T]":
+        if not self._has_built:
+            args = list(self._args)
+            args.insert(index, module)
+            self._args = tuple(args)
+            self.__construct__()
+
+        else:
+            super().insert(index, module)
+
+        return self
```
revert in favor of solution in develop

BenjaminMidtvedt commented on deeplay/models/RNN.py:
lower case module name

BenjaminMidtvedt commented on deeplay/models/__init__.py:
```diff
@@ -0,0 +1 @@
+from .RNN import *
```
imports from .py files should be explicit

BenjaminMidtvedt commented on deeplay/components/rnn.py:
```diff
@@ -1,147 +1,128 @@
-    - bidirectional (bool): If `True`, becomes a bidirectional RNN (Default: True)
-    - Embedding (nn.Embedding): Embedding layer for the input (Default: None)
+
+class Drop(nn.Module):
+    def __init__(self, p=0):
+        super(Drop, self).__init__()
+        self.p = p
+
+    def forward(self, x):
+        return nn.Dropout(p=self.p)(x[0]),x[1]
```
Rename to make clear that this is dropout for recurrent networks

BenjaminMidtvedt commented on deeplay/components/rnn.py:
```diff
@@ -1,147 +1,128 @@
-    - bidirectional (bool): If `True`, becomes a bidirectional RNN (Default: True)
-    - Embedding (nn.Embedding): Embedding layer for the input (Default: None)
+
+class Drop(nn.Module):
+    def __init__(self, p=0):
+        super(Drop, self).__init__()
+        self.p = p
+
+    def forward(self, x):
+        return nn.Dropout(p=self.p)(x[0]),x[1]
```
Use nn.functional.dropout if possible.

Henrik-KM replied:
Why is this preferred? NN.Dropout automatically disables dropout in model evaluation mode, for instance

BenjaminMidtvedt replied:
This implementation will not be disabled in eval mode, because it's not attached to the parent module. In that case, one should create the module in the init method

BenjaminMidtvedt commented on deeplay/components/rnn.py:
What's the difference between this RNN and the one in models/rnn.py?

Henrik-KM replied:
This rnn is a generic component, i.e. it needs to be configured to work for any real application. The RNN in models/rnn uses this component to configure a directly usable network given the input parameters.

Henrik-KM committed:
small bugfix (#47)

Henrik-KM committed:
PR Updates

giovannivolpe committed:
u

cmanzo commented:
@JesusPinedaC @Henrik-KM 
I'm having some trouble with the following code:

```
import deeplay as dl
from torch.nn import CrossEntropyLoss

rnn = dl.RecurrentNeuralNetwork(
    in_features=n_features,
    hidden_features=[100],
    out_features=model_number,
)
rnn.blocks.dropout.configure(p=0.2)

model_rnn = dl.Classifier(
    model=rnn,
    loss=CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
)
model_rnn.create()

trainer_rnn = dl.Trainer(max_epochs=30, accelerator="auto")
trainer_rnn.fit(model_rnn, train_loader)
```
I get `MisconfigurationException: Unknown configuration for model optimizers.` Am I doing anything wrong?

JesusPinedaC commented:
> @JesusPinedaC @Henrik-KM I'm having some trouble with the following code:
> 
> ```
> import deeplay as dl
> from torch.nn import CrossEntropyLoss
> 
> rnn = dl.RecurrentNeuralNetwork(
>     in_features=n_features,
>     hidden_features=[100],
>     out_features=model_number,
> )
> rnn.blocks.dropout.configure(p=0.2)
> 
> model_rnn = dl.Classifier(
>     model=rnn,
>     loss=CrossEntropyLoss(),
>     optimizer=dl.Adam(lr=0.001),
> )
> model_rnn.create()
> 
> trainer_rnn = dl.Trainer(max_epochs=30, accelerator="auto")
> trainer_rnn.fit(model_rnn, train_loader)
> ```
> 
> I get `MisconfigurationException: Unknown configuration for model optimizers.` Am I doing anything wrong?

@cmanzo Doing

```python
model_rnn = dl.Classifier(
    model=rnn,
    loss=CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
).create()
```
instead, solves the issue!

cmanzo commented:
It seems that:
```
rnn = dl.RNN(
    in_features=4,
    hidden_features=[100],
    out_features=5,
    rnn_type="RNN",
    dropout=0.2,
)
```
doesn't set the dropout correctly. 

However, 
`rnn.blocks.dropout.configure(p=0.2)`
does work.

There is also some inconsistency when printing the models. For example
`print(rnn)` displays:

```
RNN(
  (blocks): LayerList(
    (0): RecurrentNeuralNetwork(
  ...
          (dropout): Layer[RecurrentDropout](p=0.2)
        )
...
    (1): MultiLayerPerceptron(
...
          (dropout): Layer[Dropout](p=0.2)
        )
```

but 

```
model_rnn = dl.Classifier(
    model=rnn,
    loss=torch.nn.CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
).create()
print(model_rnn.model)
```

produces:

```
RNN(
  (blocks): LayerList(
    (0): RecurrentNeuralNetwork(
      (blocks): LayerList(
        ...
          (dropout): RecurrentDropout()
        )
 ...
      (blocks): LayerList(
        (0): LayerActivationNormalizationDropout(
 ...
          (dropout): Dropout(p=0.2, inplace=False)
        )
      )
```
However, the value is set correctly as it can be verified by:

`model_rnn.model.blocks[0].blocks[0].dropout.p`

Henrik-KM committed:
Updated RNN

Moved RNN to Applications, added RecurrentBlock. RNN blocks are no longer nested.

Henrik-KM committed:
Update rnn.py

Henrik-KM committed:
Revert "Update rnn.py"

This reverts commit b1d02b8900a5486aa9ee08b012de415fc5eac5be.

Henrik-KM committed:
Revert "Updated RNN"

This reverts commit 7fd592247455fd820f16e3291f30b8b94e1f8874.

Henrik-KM committed:
Updated RNN (#58)

* Updated RNN

Moved RNN to Applications, added RecurrentBlock. RNN blocks are no longer nested.

* Moved RNN back to models, renamed to RecurrentModel

BenjaminMidtvedt commented:
@JesusPinedaC @Henrik-KM can you check the merge conflicts?

Henrik-KM commented:
> @JesusPinedaC @Henrik-KM can you check the merge conflicts?

I think this PR can be closed, right @JesusPinedaC? All the changes should have been implemented in the separate RNN PR that was merged a few days ago.

JesusPinedaC commented:
> > @JesusPinedaC @Henrik-KM can you check the merge conflicts?
> 
> I think this PR can be closed, right @JesusPinedaC? All the changes should have been implemented in the separate RNN PR that was merged a few days ago.

I think it was not merged into develop @Henrik-KM

JesusPinedaC committed:
Merge branch 'develop' into jp/rnns

JesusPinedaC commented:
@Henrik-KM @BenjaminMidtvedt  Should we add unittests to test the new model?

BenjaminMidtvedt commented:
@JesusPinedaC yes. Should formalize exactly the things to test but check the tests by carlo as a reference

Henrik-KM committed:
Update __init__.py
